group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java

```diff
@@ -2427,25 +2471,259 @@ private boolean hasMemberSubscriptionChanged(
-                if (updatedMember.subscribedTopicRegex() != null) {
+    private static boolean isNotEmpty(String value) {
+        return value != null && !value.isEmpty();
+    }
+
+    /**
+     * Check whether the member has updated it subscribed topic regular expression and
+     * may trigger the resolution/the refresh of all the regular expressions in the
+     * group. We align the refreshment of the regular expression in order to have
+     * them trigger only one rebalance per update.
+     *
+     * @param group         The consumer group.
+     * @param member        The old member.
+     * @param updatedMember The new member.
+     * @param records       The records accumulator.
+     * @return Whether a rebalance must be triggered.
+     */
+    private boolean maybeUpdateRegularExpressions(
+        ConsumerGroup group,
+        ConsumerGroupMember member,
+        ConsumerGroupMember updatedMember,
+        List<CoordinatorRecord> records
+    ) {
+        String groupId = group.groupId();
+        String memberId = updatedMember.memberId();
+        String oldSubscribedTopicRegex = member.subscribedTopicRegex();
+        String newSubscribedTopicRegex = updatedMember.subscribedTopicRegex();
+
+        boolean bumpGroupEpoch = false;
+        boolean requireRefresh = false;
+
+        if (!Objects.equals(oldSubscribedTopicRegex, newSubscribedTopicRegex)) {
+            log.debug("[GroupId {}] Member {} updated its subscribed regex to: {}.",
+                groupId, memberId, newSubscribedTopicRegex);
+
+            if (isNotEmpty(oldSubscribedTopicRegex) && group.numSubscribedMembers(oldSubscribedTopicRegex) == 1) {
+                // If the member was the last one subscribed to the regex, we delete the
+                // resolved regular expression.
+                records.add(GroupCoordinatorRecordHelpers.newConsumerGroupRegularExpressionTombstone(
+                    groupId,
+                    oldSubscribedTopicRegex
+                ));
+            }
+
+            if (isNotEmpty(newSubscribedTopicRegex)) {
+                if (group.numSubscribedMembers(newSubscribedTopicRegex) == 0) {
+                    // If the member subscribed to a new regex, we compile it to ensure its validity.
+                    // We also trigger a refresh of the regexes in order to resolve it.
                     throwIfRegularExpressionIsInvalid(updatedMember.subscribedTopicRegex());
+                    requireRefresh = true;
+                } else {
+                    // If the new regex is already resolved, we trigger a rebalance
+                    // by bumping the group epoch.
+                    bumpGroupEpoch = group.resolvedRegularExpression(newSubscribedTopicRegex).isPresent();
                 }
-                return true;
             }
         }
-        return false;
+
+        String key = group.groupId() + "-regex";
+
+        // If a refresh is already inflight, we can schedule a new one so
```

Comment:
is this "already inflight, we don't schedule a new one"?

also, is there any concern regarding the number of epoch bumps (and rebalances) we might do? if either the executor cannot make progress (due to other group regex resolutions) or it's stuck in the event accumulator.

since we're using the defaultWriteTimeout for the runtime write event so it seems there is an upper limit inside the runtime.

Reply:
sorry for the typo. we cannot schedule a new task while one is already inflight.

i don’t get your second point. if the executor does not progress, there is no rebalances because we bump the epoch at the end.

the write timeout puts an upper limit on the write. it is not really useful here because we don’t rely on the event future.

Reply:
Ah, sorry for the confusion.

> if the executor does not progress, there is no rebalances because we bump the epoch at the end.

We bump when the regex is resolved. If it's stuck, there is no bump. Thanks that makes sense